lib/kino_flame/runner_cell.ex

```diff
@@ -126,6 +189,42 @@ defmodule KinoFLAME.RunnerCell do
 
     env = {:%{}, [], envs}
 
+    backend_ast =
+      quote do
+        {FLAME.FlyBackend,
+         [
+           unquote_splicing(specs_opts),
+           env: unquote(env)
+         ]}
+      end
+
+    to_quoted_pool(attrs, backend_ast)
+  end
+
+  defp to_quoted(%{"backend" => "k8s"} = attrs) do
+    runner_pod_tpl_a_sigil_y =
+      {:sigil_y, [delimiter: "\"\"\""],
+       [{:<<>>, [indentation: 2], [attrs["runner_pod_tpl"]]}, []]}
+
+    backend_ast = quote do: {FLAMEK8sBackend, runner_pod_tpl: runner_pod_tpl}
+    pool_ast = to_quoted_pool(attrs, backend_ast)
+
+    quote do
+      if match?({:error, _}, Code.ensure_loaded(FLAMEK8sBackend)) do
+        raise "You need to install `flame_k8s_backend` and `YamlElixir` as dependency."
+      end
+
+      import YamlElixir.Sigil
+
+      runner_pod_tpl =
+        unquote(runner_pod_tpl_a_sigil_y)
+        |> KinoFLAME.RunnerCell.add_livebook_cookie()
```

Comment:
Smart cell modules are private, so we should not delegate work to this function. But more broadly speaking, the code that smart cell generates should reflect what the user would ideally write, so it should not depend on custom functions like this. Instead, we want to modify the template upfront and inline the modified template here :)

Reply:
Okay. In order to do that, I will need to add `yaml_elixir` and `ymlr` to the list of dependencies of `kino_flame`, though.

Reply:
In that case, we can make them optional dependencies, we check if they are loaded, if not we generate empty code. And then we have the warning for missing deps, as per the other comment.

Reply:
I assume the flame_k8s_backend do not depend on them? Which I guess it makes sense, we want to keep backend dependencies to a minimum. I will take a look at this PR and see if we can get rid of yaml dependency.

Reply:
Ok, I think we can get rid of YAMLElixir and, instead of adding the cookie dynamically, just hardcode that into an Elixir sigil. So for example, instead of:

```
~Y"""
i_dont_know:
  k8s
metadata:
  something: ...
"""
|> add_livebook_cookie()
```

We will do this:

```
"""
i_dont_know:
  k8s
metadata:
  env:
    LIVEBOOK_COOKIE: #{Node.get_cookie()}
"""
```

The downside is that users will see `LIVEBOOK_COOKIE: "#{Node.get_cookie()}"` but I think that's fine. If we are worried about potential conflicts, we could also use EEx to evaluate it :)

That should simplify things. WDYT?

Reply:
Oh, the other thing is that we lose YAML validation if we don't use `~Y` sigil, but we can probably add YAML linting to CodeMirror. Is that possible @jonatanklosko?

Reply:
Well... at the moment, the K8s backend expects a map. the `~y` sigil actually parses the string into a map. Without `yaml_elixir`, we can't do that...

Reply:
@josevalim my understanding is that we need `yaml_elixir` regardless, because the backend expects the parsed Elixir terms.

Sidenote: from a quick search I didn't find a drop-in linter, we could implement one, but that's additional JS dependency for YAML, so I would be a bit hesitant. Though maybe we could lazyload it only when using the YAML language.

Reply:
Oh but then it's only needed on the runtime, not as a dep to `kino_flame`.
But I have two questions: 

* How on earth do I modify a string if I can't parse it first? I mean... the list of env can be anywhere in there!
* How can I create an AST of a multiline binary?

Reply:
I think it is okay to require dependencies on the runtime (i.e. if the generated code uses `~y` sigil, that is okay. BUT we don't want more deps on `kino_flame`, especially since they are only needed for Kubernetes backned. 

So actually, we can "just" inject the env into the generated string. My question is: How do we do that without parsing the yaml first?

Reply:
> we don't want more deps on kino_flame

If a dependency is required on the runtime, then I think it's fine to depend on it in `kino_flame` as an optional dependency. `kino_flame` would need it only when the Kubernetes backend is selected, which is exactly when the runtime would need it also.

Sidenote: we should have `flame_k8s_backend` as an optional dependency also. The reasoning is that we want some kind of requirement on the `:flame_k8s_backend` version, because we generate code that uses the backend, and so we want to make sure the code is compatible. We do the same in [`kino_db`](https://github.com/livebook-dev/kino_db/blob/bb6af5bc11fec657cc76c974e6684ae48ac314ae/mix.exs#L31-L41).

`ymlr` is a different story though, because it's not required by the end user, so ideally we wouldn't relay on it just for the smart cell.

Reply:
Sorry, I still don't see a viable solution here. The problem we'd like to solve: add an env variable to the pod template. At the same time, in the generated code, I'd like to show YAML because that's what the user entered. Maybe I didn't see the solution in the comments, but... the only way I currently see is: `yaml -> map -> add_the_env -> yaml`. And for `yaml -> map` I need `yaml_elixir` and for `map -> yaml` I need `ymlr`. 
So... what about `ymlr`?

Reply:
At the risk of losing focus - there are complete different ways we could design the smart cell for k8s backend (we had similar discussions for the k8s runtime):

* offer a form, similar to the Fly backend. The problem with this is: If you want to use FLAME for more than just trying out, you want to be able to configure the pod template beyond what's possible through the form.
* Let users select the template from a list of `PodTemplate` resources we retrieve from the cluster. But this makes the smart cell much more difficult to use as users would have to apply pod templates to the cluster first...

Reply:
There are three options:

1. Use map instead of yaml in the generated code. The weird part is that then `yaml_elixir` is no longer runtime dependency, but we need it. But at the end of the day I would fine having the smart cell depend on that, even non-optionally.

2. Wrap the user yaml in sigil as is and inline simples possible code to add the cookie, like this:

   ```elixir
   ~y"""
   ...
   """
   |> update_in(
     ["spec", "containers", Access.at(0), "env"],
     &[%{"name" => "LIVEBOOK_COOKIE", "value" => Node.get_cookie()} | &1]
   )
   ```

3. I think what Jose meant is to put `LIVEBOOK_COOKIE: "#{Node.get_cookie()}"` in the initial template, so when we wrap it in the sigil, we use actual Elixir interpolation. Or we could support syntax like `{{elixir code}}` and rewrite occurrences to Elixir interpolation. The main downside is that we relay on the user to keep that env in the template, otherwise things just don't work.

Is there a specific reason we want the smart cell to use `~y` instead of Elixir terms? Is it because without the cell, the user would most likely use YamlElixir anyway?

Reply:
>  think what Jose meant is to put LIVEBOOK_COOKIE: "#{Node.get_cookie()}" in the initial template, so when we wrap it in the sigil, we use actual Elixir interpolation. 

ooooh now I get it! So we depend on users putting that there (even if they replace the whole template). Okay that would be doable.

>  The main downside is that we relay on the user to keep that env in the template, otherwise things just don't work.
If you're find with `yaml_elixir`as a dep, we _could_ test for its existence :)

> Is there a specific reason we want the smart cell to use ~y instead of Elixir terms? Is it because without the cell, the user would most likely use YamlElixir anyway?

With "Elixir terms" you mean a map? It's just that with the Sigil, the code looks like what the user entered. 
Or are you talking about `~y` vs. `|> YamlElixir.read_from_string!()`? If you're talking about this, honestly, I did not figure out how to create the AST for a multiline string without a sigil 🙈

Reply:
> If you're fine with having yaml_elixiras a dep, we could test for its existence :)

Oh, actually we could as well check if `=~ "LIVEBOOK_COOKIE` and warn otherwise, telling them what to put.

> With "Elixir terms" you mean a map? It's just that with the Sigil, the code looks like what the user entered.

Yeah, a map. I don't think it necessarily needs to look like the user entered, because here we effectively treat YAML as a flexible alternative to form. The question is rather, if the user didn't use the smart cell, do we expect them to write `~y` or maps by hand?

> I did not figure out how to create the AST for a multiline string without a sigil 🙈

Just for the record, you can do that by specifying delimiter and adding newline to the string: `{:sigil_S, [delimiter: ~S["""]], [{:<<>>, [], [string <> "\n"]}, []]}` :)

Reply:
> Oh, actually we could as well check if =~ "LIVEBOOK_COOKIE and warn otherwise, telling them what to put.

True.

> The question is rather, if the user didn't use the smart cell, do we expect them to write ~y or maps by hand?

Yes I'd expect them to use the `~y` sigil which probably most users don't know about so my idea was also using the smart cell to show users this sigil. :)

> Just for the record...

oh... `:sigil_S`! True! :)

Reply:
Actually.... about the blob of yaml: In `flame_k8s_backend` until now I was only depending on `flame`. But now I have to add `yaml_elixir` there. Wouldn't it be better to require that dependency on the user's side?

Reply:
I agree we want to avoid the dependency on `flame_k8s_backend`, as that has other implications. So I think, overall, we either have to ask the users to write JSON or have `yaml_elixir` as a dependency. I think the `yaml_elixir` dependency is fine, we just need to include it in Livebook too. I think we can ship it.